storage/SMAIndexSubBlock.cpp

```diff
@@ -0,0 +1,665 @@
+  entries_ = reinterpret_cast<SMAEntry*>(header_ + 1);
+  attribute_to_entry_ = new int[total_attributes_];
+  // Set each attribute to invalid.
+  for (std::size_t i = 0; i < total_attributes_; ++i) {
+    attribute_to_entry_[i] = -1;
+  }
+
+  if (new_block) {
+    header_->index_consistent = false;
+  }
+
+  // Zero the comparator arrays so we don't try to reference invalid memory later.
+  memset(less_comparisons_, 0, sizeof(UncheckedComparator*) * kNumTypeIDs);
+  memset(equal_comparisons_, 0, sizeof(UncheckedComparator*) * kNumTypeIDs);
+  memset(add_operations_, 0, sizeof(UncheckedBinaryOperator*) * kNumTypeIDs);
+
+  // Iterate through each attribute which is being indexed.
+  for (std::size_t indexed_attribute_num = 0;
+       indexed_attribute_num < indexed_attributes_;
+       ++indexed_attribute_num) {
+    const attribute_id attribute = description_.GetExtension(
+        SMAIndexSubBlockDescription::indexed_attribute_id,
+        indexed_attribute_num);
+
+    const Type &attribute_type = tuple_store_.getRelation().getAttributeById(attribute)->getType();
+    SMAEntry *entry = entries_ + indexed_attribute_num;
+
+    // Initialize the operator map.
+    if (sma_internal::canSum(attribute_type.getTypeID())) {
+      TypeID attr_typeid = attribute_type.getTypeID();
+      TypeID attr_sum_typeid = sma_internal::sumType(attr_typeid);
+      if (add_operations_[attr_typeid] == nullptr) {
+        add_operations_[attr_typeid]
+          = BinaryOperationFactory::GetBinaryOperation(BinaryOperationID::kAdd)
+              .makeUncheckedBinaryOperatorForTypes(TypeFactory::GetType(attr_typeid),
+                                                   TypeFactory::GetType(attr_sum_typeid));
+      }
+    }
+
+    // Initialize comparison map. Maps attribute type to comparison function.
+    int attr_typeid = static_cast<int>(attribute_type.getTypeID());
+    if (less_comparisons_[attr_typeid] == nullptr) {
+      DCHECK(equal_comparisons_[attr_typeid] == nullptr);
+      less_comparisons_[attr_typeid]
+          = ComparisonFactory::GetComparison(ComparisonID::kLess)
+              .makeUncheckedComparatorForTypes(attribute_type, attribute_type);
+      equal_comparisons_[attr_typeid]
+          = ComparisonFactory::GetComparison(ComparisonID::kEqual)
+              .makeUncheckedComparatorForTypes(attribute_type, attribute_type);
+    }
+
+    // Map the attribute's id to its entry's index.
+    attribute_to_entry_[attribute] = indexed_attribute_num;
+
+    if (!header_->index_consistent) {
+      resetEntry(entry, attribute, attribute_type);
+    } else {
+      // If the data held by min/max is out of line, we must retrieve it as we initialize.
+      if (!TypedValue::RepresentedInline(entry->type_id)) {
+        // First, set to 0 so that we don't try to free invalid memory on copy.
+        // Next, copy from the tuple store. This will copy and give us ownership of out of line data.
```

Comment:
I am confused by the usage of copying and getting ownership in the same context. Can you elaborate?